api/envoy/config/trace/v3/opentelemetry.proto

```diff
@@ -17,10 +21,41 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 // Configuration for the OpenTelemetry tracer.
 //  [#extension: envoy.tracers.opentelemetry]
 message OpenTelemetryConfig {
-  // The upstream gRPC cluster that will receive OTLP traces.
-  // Note that the tracer drops traces if the server does not read data fast enough.
-  // This field can be left empty to disable reporting traces to the collector.
-  core.v3.GrpcService grpc_service = 1;
+  // Configuration for sending traces to the collector over HTTP. Note that
+  // this will only be used if the grpc_service is not set above.
+  message HttpConfig {
+
+    // The upstream cluster that will receive OTLP traces over HTTP.
+    string cluster_name = 1;
+
+    // The path to the trace ingest endpoint. The path is appended with the host name configured in the cluster.
+    // Optional: If omitted, "/v1/traces" will be used as default.
+    string traces_path = 2;
+
+    // Sets the maximum duration in milliseconds that a response can take to arrive upon request.
+    google.protobuf.Duration timeout = 3 [(validate.rules).duration = {
+      required: true
+      gte {}
+    }];
+
+    // Custom header values to be added to the OTLP HTTP request.
+    repeated config.core.v3.HeaderValue headers = 4;
+
+    // The hostname to include in the Host header of the OTLP HTTP request.
+    string hostname = 5;
+  }
+
+  oneof export_protocol {
```

Comment:
@AlexanderEllis @htuch I need your input in this change:

I wanted to split the configuration here, so HTTP things are not "mixed" with gRPC. I managed to make it work, but that required me to create a new Factory ([config.cc](https://github.com/envoyproxy/envoy/blob/main/source/extensions/tracers/opentelemetry/config.cc#L16)) and register it under a different name, since it can't have the same. For ex `envoy.tracers.opentelemetry.http`. That worked but I don't think it's a good idea. It seems like a total different tracer but it's not, just a new exporter so I abandoned the idea. 


I still wanted to make it obvious that HTTP config cannot be mixed with gRPC so I thought of using `oneof` here to make it more explicit. The problem is that this introduces a change where one of them is required. There is a test that checks that you can configure the tracer but not populate the `grpc_service`, essentially "disabling" reporting spans. Now with this change, if you leave one of them empty it crashes and envoy does not start. (I modified the test)

My question now to you: Do you think this new behavior is breaking? To me, it doesn't make sense to configure the tracer but leave the rest empty to "disable" it, but maybe there is some prior-art/need for this that I'm not aware. 

I also wanted to add some custom validation in code but I omitted so far: If both are configured fail to prevent strange behavior. I think now with `oneof` the last that is configured "wins". I think adding validation and failing fast removes possible mis-configuration. Users may think they can export to both channels simultaneously which is not the case. 

Sorry for the long text but I think this is an important point we need to discuss.

Reply:
In general, adding `oneof`` to an existing field is considered a breaking change, see https://github.com/envoyproxy/envoy/blob/main/api/API_VERSIONING.md. 

You can achieve the same effect by documentation and validating mutex in code on config ingestion.

@adisuissa do you think we should have a general `HttpService` to complement the existing `GrpcService`?

Reply:
> In general, adding `oneof`` to an existing field is considered a breaking change, see [main/api/API_VERSIONING.md](https://github.com/envoyproxy/envoy/blob/main/api/API_VERSIONING.md?rgh-link-date=2023-08-23T14%3A59%3A09Z).

Ah ok I missed that API document, but it aligns with what I suspected. 

> validating mutex in code on config ingestion.

Sorry, not sure I fully understand it. Do you mean removing the `oneof` and simply checking in code if both `GrpcService` and `HttpConfig` are both populated? I was thinking on doing this initially and maybe throwing if that's the case. 

>do you think we should have a general HttpService to complement the existing GrpcService

I was also thinking the same, `HttpConfig` makes sense, but it looks nicer if it matches what we have tiwht `GrpcService`. I can change if you think that is better :)

Reply:
@htuch I reverted the `OneOf`, renamed `http_config` to `http_service` to match `grpc_service`. 

I still think it makes sense to validate the case where the user configures both the gRPC and HTTP exporters now. What I did in c3e8b819f08c60bf5b3a0e34878b5eb27cf7547b is I throw an exception when building the OpenTelemetry tracer. 

In OTel, this I think aligns with the spec wording on "SDK initialization" https://github.com/open-telemetry/opentelemetry-specification/blob/v1.24.0/specification/error-handling.md?plain=1#L18

I also read this in Envoy https://github.com/envoyproxy/envoy/blob/main/STYLE.md#error-handling and I think it also aligns with :

> Although we strongly recommend that any type of startup error leads to a fatal error, since this is almost always a result of faulty configuration which should be caught during a canary process

let me know what you think. I adapted the tests for this, and also added a new one to cover the case when both exporters are configured.

Reply:
Exceptions are fine, but need to be done at config ingestion time. Still interested in whether we should have a core/base notion of `HttpService` from @adisuissa and/or @envoyproxy/api-shepherds.

Reply:
Apologies for the delay, I was OOO.

> @adisuissa do you think we should have a general `HttpService` to complement the existing `GrpcService`? ext_authz is another example of this pattern. There's usually somethign extension specific and some common aspects.

For the common part, I'm assuming you are referring to things like retries, timeouts, and such. If that's the case, then yeah, it may make sense to either include them with the `cluster_name` in an `HttpService` kind of type, or if they are a property of the cluster itself, then directly in the cluster type. I'm assuming the former is better as one may want to configure different timeouts to the cluster, depending on the route.
For this PR, I suggest making a new type `HttpService` in "api/envoy/config/core/v3/http_service.proto", that for now can have just the `cluster_name`, but later can grow with timeouts and retries policies and such.

Reply:
> For this PR, I suggest making a new type HttpService in "api/envoy/config/core/v3/http_service.proto", that for now can have just the cluster_name, but later can grow with timeouts and retries policies and such.

@adisuissa so the new type only has the cluster name? And what about the rest of the config? For the HTTP exporter we need now:

1. cluster name
2. path for the trace endpoint
3. host name
4. headers
5. timeout 

1, 3, 4 and 5 I think can be part of the new `HttpService` type. Not sure about 2 though. 

The timeout is used today when sending the HTTP request. Not sure if it's really needed or if we can omit it. The GrpcService also has it today which maybe can be used as an argument the new type should also have it.

Reply:
That's a good question. I think that the cluster-name and path are the minimum to define a service (similar to a gRPC one). In this case the path is for the OLTP traces, but generally speaking it could be a path to any service.

The hostname, custom headers, and timeout seem to be more of a route property, but considering that the path is static, then they may as well be part of the HttpService. I suggest following the naming conventions (and comments) as in https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/route/v3/route_components.proto, to keep consistency between the two. Also note that custom-headers can be both on the request path and maybe on the response path (no need to add response-path support, just to clarify that this is only for requests).

@htuch for thoughts on this.

Reply:
I introduced the new type in 6815fb5ade2987d79dc0620d0c30877b88b8b612. Let me know what you think!